### DISCUSSION TITLE
Adapting a single migration to include data creation/manipulation

### DISCUSSION THREAD

**Discussion by hardysabs2** [prisma/prisma#21479] [Q&A]
### Question

It seems to me that the single auto seed lends itself more to first time building of schema and data from the ground up.
However, once a project schema and its data is reasonably well established in a staging or prod environment I would think we want to JUST* run incremental schema changes.  Do we do this with:
Q1: prisma migrate deploy OR
prisma migrate dev ?
Q2: what exactly is the difference ?

*Having said this, there are often valid reasons during the life of the project to need to create or manipulate data ONCE at the point of a given migration.  A few examples:
a) Introducing a missing FK relationship may require tidy up, deleting orphans etc before running the migration SQL
b) After the creation of a utility table one may want to populate it with data
c) introduction of a lookup table where the requirements change to a many to many relationship, one will want to run a script to initially populate the look up for the FK/PK values of the original table.

Q3: I have seen posts about adding one off seeds to migrations (2021) but do not see any progress there.  Has there been progress there?

Of course often the need is more complex than straight forward insertion.  
I understand migrations are generated from a diff of the old and new static ORM schema.  
Q4: Is there any reason we are not at liberty to edit a migration once it has been generated to add further essential SQL so long as we properly document it as a manual edit.
Q5: If the complexity lent itself better to a script, is there a way to run a script to do the same task ONCE just at the point of the migration?

Thank you so much, I would really appreciate your input on these questions (Qn:) above :-)

Andrew

**Comment by ludralph** [comment]
Hi @hardysabs2 👋 

Thank you for raising this question.

> Q1: prisma migrate deploy OR prisma migrate dev ?
   Q2: what exactly is the difference ?

The difference between `prisma migrate deploy` and `prisma migrate dev` lies in their intended use cases. `prisma migrate dev` is used for creating migrations and applying them during local development. It's a development command that resets your database, applies migrations and generates artifacts (e.g., Prisma Client). On the other hand, `prisma migrate deploy` is used for applying generated migrations to production. It's a command that should be used when you want to apply migrations in a production environment. It does not reset the database and does not generate artifacts. [reference](https://www.prisma.io/docs/concepts/components/prisma-migrate/migrate-development-production)

> Q3: I have seen posts about adding one off seeds to migrations (2021) but do not see any progress there. Has there been progress there?

There is an open [GitHub issue](https://github.com/prisma/prisma/issues/13290) discussing the possibility of running migrations only inside specific contexts, which could potentially be used for one-off seeds to migrations. Is that the post you were referring to? However, this has not been prioritized yet.

> Q4: Is there any reason we are not at liberty to edit a migration once it has been generated to add further essential SQL so long as we properly document it as a manual edit.

You can edit the migrations to store any SQL statement you want. You will use the `--create-only` flag and then add the changes you want. [Read more](https://www.prisma.io/docs/guides/database/developing-with-prisma-migrate/customizing-migrations.)

> Q5: If the complexity lent itself better to a script, is there a way to run a script to do the same task ONCE just at the point of the migration?

Prisma Migrate does support seeding, which allows you to bootstrap a usable test environment with data quickly. Seeding is currently supported via scripts written in TypeScript, JavaScript, and Shell. To use seeding, you define the script inside the prisma folder and run the prisma db seed command. Learn more on seeding with prisma migrate [here](https://www.prisma.io/docs/guides/migrate/seed-database#integrated-seeding-with-prisma-migrate)

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
